File: src/java/jpdp/core/Lexicon.java
/*
 * Copyright (c) Konrad Grzanek. All rights reserved.
 * Created 2015-11-18
 */
package jpdp.core;

import java.util.HashMap;
import java.util.Map;

public final class Lexicon {

  public Prop getProp(Object name) {
    return propsCache.get(name.toString());
  }

  public Value getValue(Object origin) {
    return valuesCache.get(origin);
  }

  Lexicon(Iterable<Object> propNames, Iterable<Object> valueOrigins) {
    this.propsCache = makePropsCache(propNames);
    this.valuesCache = makeValuesCache(valueOrigins);
  }

  private final Map<String, Prop> propsCache;

  private final Map<Object, Value> valuesCache;

  private static Map<String, Prop> makePropsCache(Iterable<Object> names) {
    final Map<String, Prop> cache = new HashMap<>();
    int code = 0;
    for (Object name : names) {
      String s = name.toString();
      if (cache.containsKey(s)) {
        throw new IllegalArgumentException("Duplicate property " + s);
      }
      cache.put(s, new Prop(s, code));
      code += 1;
    }
    return cache;
  }

  private static Map<Object, Value> makeValuesCache(Iterable<Object> origins) {
    final Map<Object, Value> cache = new HashMap<>();
    for (Object origin : origins) {
      if (cache.containsKey(origin)) {
        throw new IllegalArgumentException("Duplicate value origin " + origin);
      }
      cache.put(origin, new Value(origin));
    }
    return cache;
  }

}

File: src/clj/pdp/core/lexicon.clj
;; Copyright (c) Konrad Grzanek. All rights reserved.
;; Created 2015-11-18

(in-ns 'pdp.core)

(defmacro defprops
  [& names]
  (when-not (= names (distinct names))
    (terror "Prop names must be distinct."))

  (let [inames  (map pair names (iterate inc 0))
        clause  (fn [[name code]]
                  `(defl ~name (jpdp.core.Prop. ~(str name) ~(int code))))
        clauses (map clause inames)]
    `(do ~@clauses)))


(defmacro defvals
  [& origins]
  (when-not (= origins (distinct origins))
    (terror "Origins of Values must be distinct."))

  `(do ~@(map (fn [origin]
                `(defl ~origin (jpdp.core.Value. ~(str origin))))
              origins)))
